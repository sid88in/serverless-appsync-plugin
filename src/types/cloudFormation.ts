import { AWS } from '@serverless/typescript';

export type FnGetAtt = {
  'Fn::GetAtt': string[];
};

export type FnJoin = {
  'Fn::Join': [string, (string | IntrinsicFunction)[]];
};

export type FnRef = {
  Ref: string;
};

export type FnSub = {
  'Fn::Sub': [string, Record<string, string | IntrinsicFunction>];
};

export type IntrinsicFunction = FnGetAtt | FnJoin | FnRef | FnSub;

export type CfnDeltaSyncConfig = {
  BaseTableTTL: number;
  DeltaSyncTableName: string | IntrinsicFunction;
  DeltaSyncTableTTL: number;
};

export type CfnResources = Required<Required<AWS>['resources']>['Resources'];
export type CfnResource = CfnResources[string];

export type CfnDataSource = {
  Type: 'AWS::AppSync::DataSource';
  Properties: {
    ApiId: string | IntrinsicFunction;
    Name: string | IntrinsicFunction;
    Description?: string;
    Type?:
      | 'AWS_LAMBDA'
      | 'AMAZON_DYNAMODB'
      | 'AMAZON_OPENSEARCH_SERVICE'
      | 'NONE'
      | 'HTTP'
      | 'RELATIONAL_DATABASE';
    ServiceRoleArn?: string | IntrinsicFunction;
    LambdaConfig?: {
      LambdaFunctionArn: string | IntrinsicFunction;
    };
    DynamoDBConfig?: {
      TableName: string | IntrinsicFunction;
      AwsRegion: string | IntrinsicFunction;
      UseCallerCredentials: boolean;
      Versioned?: boolean;
      DeltaSyncConfig?: CfnDeltaSyncConfig;
    };
    OpenSearchServiceConfig?: {
      AwsRegion: string | IntrinsicFunction;
      Endpoint: string | IntrinsicFunction;
    };
    RelationalDatabaseConfig?: {
      RelationalDatabaseSourceType: 'RDS_HTTP_ENDPOINT' | IntrinsicFunction;
      RdsHttpEndpointConfig?: {
        AwsRegion: string | IntrinsicFunction;
        DbClusterIdentifier: string | IntrinsicFunction;
        DatabaseName?: string | IntrinsicFunction;
        Schema?: string | IntrinsicFunction;
        AwsSecretStoreArn: string | IntrinsicFunction;
      };
    };
    HttpConfig?: {
      Endpoint: string | IntrinsicFunction;
      AuthorizationConfig?: {
        AuthorizationType?: string;
        AwsIamConfig?: {
          SigningRegion: string | IntrinsicFunction;
          SigningServiceName?: string | IntrinsicFunction;
        };
      };
    };
  };
};

export type CfnAppSyncRuntime = {
  Name: 'APPSYNC_JS';
  RuntimeVersion: '1.0.0';
};

export type CfnResolver = {
  RequestMappingTemplate?: string;
  ResponseMappingTemplate?: string;
  Properties: {
    ApiId: string | IntrinsicFunction;
    TypeName: string;
    FieldName: string;
    Kind?: 'PIPELINE' | 'UNIT';
    DataSourceName?: string | IntrinsicFunction;
    Description?: string;
    FunctionVersion?: string;
    Code?: string | IntrinsicFunction;
    Runtime?: CfnAppSyncRuntime;
    RequestMappingTemplate?: string | IntrinsicFunction;
    ResponseMappingTemplate?: string | IntrinsicFunction;
    PipelineConfig?: {
      Functions: (string | IntrinsicFunction)[];
    };
    CachingConfig?: {
      Ttl: number;
      CachingKeys?: string[];
    };
    SyncConfig?: {
      ConflictDetection: 'VERSION' | 'NONE';
      ConflictHandler?: 'OPTIMISTIC_CONCURRENCY' | 'AUTOMERGE' | 'LAMBDA';
      LambdaConflictHandlerConfig?: {
        LambdaConflictHandlerArn: string | IntrinsicFunction;
      };
    };
    MaxBatchSize?: number;
  };
};

export type CfnFunctionResolver = {
  Type: 'AWS::AppSync::FunctionConfiguration';
  Properties: {
    ApiId: string | IntrinsicFunction;
    Name: string | IntrinsicFunction;
    DataSourceName: string | IntrinsicFunction;
    Description?: string;
    FunctionVersion?: string;
    Code?: string | IntrinsicFunction;
    Runtime?: CfnAppSyncRuntime;
    RequestMappingTemplate?: string | IntrinsicFunction;
    ResponseMappingTemplate?: string | IntrinsicFunction;
    SyncConfig?: {
      ConflictDetection: 'VERSION' | 'NONE';
      ConflictHandler?: 'OPTIMISTIC_CONCURRENCY' | 'AUTOMERGE' | 'LAMBDA';
      LambdaConflictHandlerConfig?: {
        LambdaConflictHandlerArn: string | IntrinsicFunction;
      };
    };
    MaxBatchSize?: number;
  };
};

export type CfnApiKey = {
  Type: 'AWS::AppSync::ApiKey';
  Properties: {
    ApiId: string | IntrinsicFunction;
    Description?: string;
    Expires: number;
    ApiKeyId?: string;
  };
};

export type CfnWafAction = { [key: string]: Record<string, never> };

export type CfnWafRule = {
  Action?: CfnWafAction;
  Name: string;
  OverrideAction?: CfnWafAction;
  Priority?: number;
  Statement: CfnWafRuleStatement;
  VisibilityConfig: unknown;
};

export type CfnWafRuleStatement = {
  RateBasedStatement?: CfnWafRuleRateBasedStatement;
  ByteMatchStatement?: CfnWafRuleRateByteMatchStatement;
  AndStatement?: {
    Statements: CfnWafRuleStatement[];
  };
  [k: string]: Record<string, unknown> | undefined;
};

type CfnWafRuleRateBasedStatement = {
  AggregateKeyType: string;
  ForwardedIPConfig?: {
    FallbackBehavior: string;
    HeaderName: string;
  };
  Limit: number;
  ScopeDownStatement?: CfnWafRuleStatement;
};

type CfnWafRuleRateByteMatchStatement = {
  FieldToMatch: unknown;
  PositionalConstraint: string;
  SearchString?: string | IntrinsicFunction;
  SearchStringBase64?: string;
  TextTransformations: {
    Priority: number;
    Type: string;
  }[];
};

export type CfnOutput = {
  OutputKey: string;
  OutputValue: string;
};
