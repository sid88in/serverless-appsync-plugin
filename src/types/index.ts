import { BuildOptions } from 'esbuild';
import {
  ApiKeyConfig,
  Auth,
  Substitutions,
  CachingConfig,
  DomainConfig,
  LoggingConfig,
  WafConfig,
  DsDynamoDBConfig,
  DsEventBridgeConfig,
  DsHttpConfig,
  DsLambdaConfig,
  DsNone,
  DsOpenSearchConfig,
  DsRelationalDbConfig,
  SyncConfig,
  EnvironmentVariables,
} from './common';
export * from './common';

export type AppSyncConfig = {
  name: string;
  schema?: string | string[];
  authentication: Auth;
  additionalAuthentications?: Auth[];
  domain?: DomainConfig;
  apiKeys?: (ApiKeyConfig | string)[];
  resolvers?: Record<string, ResolverConfig>[] | Record<string, ResolverConfig>;
  pipelineFunctions?:
    | Record<string, PipelineFunctionConfig>[]
    | Record<string, PipelineFunctionConfig>;
  dataSources:
    | Record<string, DataSourceConfig>[]
    | Record<string, DataSourceConfig>;
  substitutions?: Substitutions;
  environment?: EnvironmentVariables;
  xrayEnabled?: boolean;
  logging?: LoggingConfig;
  caching?: CachingConfig;
  waf?: WafConfig;
  tags?: Record<string, string>;
  visibility?: 'GLOBAL' | 'PRIVATE';
  esbuild?: BuildOptions | false;
  introspection?: boolean;
  queryDepthLimit?: number;
  resolverCountLimit?: number;
};

export type BaseResolverConfig = {
  field?: string;
  type?: string;
  request?: string | false;
  response?: string | false;
  code?: string;
  caching?:
    | {
        ttl?: number;
        keys?: string[];
      }
    | boolean;
  sync?: SyncConfig;
  substitutions?: Substitutions;
};

export type ResolverConfig = UnitResolverConfig | PipelineResolverConfig;

export type UnitResolverConfig = BaseResolverConfig & {
  kind: 'UNIT';
  dataSource: string | DataSourceConfig;
  maxBatchSize?: number;
};

export type PipelineResolverConfig = BaseResolverConfig & {
  kind?: 'PIPELINE';
  functions: (string | PipelineFunctionConfig)[];
};

export type DataSourceConfig = {
  description?: string;
} & (
  | DsHttpConfig
  | DsDynamoDBConfig
  | DsRelationalDbConfig
  | DsOpenSearchConfig
  | DsLambdaConfig
  | DsEventBridgeConfig
  | DsNone
);

export type PipelineFunctionConfig = {
  dataSource: string | DataSourceConfig;
  description?: string;
  code?: string;
  request?: string;
  response?: string;
  maxBatchSize?: number;
  substitutions?: Substitutions;
  sync?: SyncConfig;
};
