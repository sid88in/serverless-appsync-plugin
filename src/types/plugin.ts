import { BuildOptions } from 'esbuild';
import {
  Auth,
  DomainConfig,
  ApiKeyConfig,
  LoggingConfig,
  CachingConfig,
  WafConfig,
  SyncConfig,
  DsHttpConfig,
  DsDynamoDBConfig,
  DsRelationalDbConfig,
  DsOpenSearchConfig,
  DsLambdaConfig,
  DsEventBridgeConfig,
  DsNone,
  Substitutions,
  EnvironmentVariables,
  EnhancedMetricsConfig,
} from './common';
export * from './common';

export type AppSyncConfig = {
  name: string;
  schema: string[];
  authentication: Auth;
  additionalAuthentications: Auth[];
  domain?: DomainConfig;
  apiKeys?: Record<string, ApiKeyConfig>;
  dataSources: Record<string, DataSourceConfig>;
  resolvers: Record<string, ResolverConfig>;
  pipelineFunctions: Record<string, PipelineFunctionConfig>;
  substitutions?: Substitutions;
  environment?: EnvironmentVariables;
  enhancedMetrics?: EnhancedMetricsConfig;
  xrayEnabled?: boolean;
  logging?: LoggingConfig;
  caching?: CachingConfig;
  waf?: WafConfig;
  tags?: Record<string, string>;
  visibility?: 'GLOBAL' | 'PRIVATE';
  esbuild?: BuildOptions | false;
  introspection?: boolean;
  queryDepthLimit?: number;
  resolverCountLimit?: number;
};

export type BaseResolverConfig = {
  field: string;
  type: string;
  request?: string | false;
  response?: string | false;
  code?: string;
  caching?:
    | {
        ttl?: number;
        keys?: string[];
      }
    | boolean;
  sync?: SyncConfig;
  substitutions?: Substitutions;
};

export type ResolverConfig = UnitResolverConfig | PipelineResolverConfig;

export type UnitResolverConfig = BaseResolverConfig & {
  kind: 'UNIT';
  dataSource: string;
  maxBatchSize?: number;
};

export type PipelineResolverConfig = BaseResolverConfig & {
  kind?: 'PIPELINE';
  functions: string[];
};

export type DataSourceConfig = {
  name: string;
  description?: string;
} & (
  | DsHttpConfig
  | DsDynamoDBConfig
  | DsRelationalDbConfig
  | DsOpenSearchConfig
  | DsLambdaConfig
  | DsEventBridgeConfig
  | DsNone
);

export type PipelineFunctionConfig = {
  name: string;
  dataSource: string;
  description?: string;
  code?: string;
  request?: string;
  response?: string;
  maxBatchSize?: number;
  substitutions?: Substitutions;
  sync?: SyncConfig;
};
export { Substitutions };
