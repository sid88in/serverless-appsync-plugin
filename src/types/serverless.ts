import type { AWS } from '@serverless/typescript';
import { AppSyncConfigInput } from '../getAppSyncConfig';

export type Hook = () => void | Promise<void>;

export type Provider = {
  naming: {
    getStackName: () => string;
    getLambdaLogicalId: (functionName: string) => string;
  };
  request: <Input, Output>(
    service: string,
    method: string,
    params: Input,
  ) => Promise<Output>;
};

export type Serverless = {
  serviceDir: string;
  config: {
    servicePath: string;
  };
  processedInput: {
    functions?: Record<string, unknown>;
    options: Record<string, unknown>;
  };
  configurationInput: AWS & {
    appSync: AppSyncConfigInput;
  };
  service: AWS & {
    setFunctionNames(rawOptions: Record<string, unknown>): void;
  };
  configSchemaHandler: {
    defineTopLevelProperty: (
      name: string,
      schema: Record<string, unknown>,
    ) => void;
  };
  getProvider: (provider: 'aws') => Provider;
  cli: {
    log: (
      message: string,
      entity?: string,
      opts?: Record<string, unknown>,
    ) => void;
  };
  // Serverless v3
  addServiceOutputSection: (
    section: string,
    content: string | string[],
  ) => void;
  classes: {
    Error: ErrorConstructor;
  };
};

export type CloudformationTemplate = AWS['resources'];

export type CommandsDefinition = Record<
  string,
  {
    lifecycleEvents?: string[];
    commands?: CommandsDefinition;
    usage?: string;
    options?: {
      [name: string]: {
        usage: string;
        required?: boolean;
        shortcut?: string;
        type?: 'string' | 'boolean' | 'multiple';
      };
    };
  }
>;

export type CliOptions = Record<string, string | boolean | string[]>;

/**
 * Serverless v3
 */

export type ServerlessLogger = {
  error: (message: string) => void;
  warning: (message: string) => void;
  notice: (message: string) => void;
  info: (message: string) => void;
  debug: (message: string) => void;
  success: (message: string) => void;
};

export type ServerlessHelpers = {
  log: ServerlessLogger;
  writeText: (message: string) => void;
};
