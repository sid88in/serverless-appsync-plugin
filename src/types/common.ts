import { CfnWafRuleStatement, IntrinsicFunction } from './cloudFormation';

export type IamStatement = {
  Effect: 'Allow' | 'Deny';
  Action: string[];
  Resource: string | IntrinsicFunction | (string | IntrinsicFunction)[];
};

export type WafConfig = {
  enabled?: boolean;
  arn?: string;
  name?: string;
  defaultAction?: WafAction;
  description?: string;
  visibilityConfig?: VisibilityConfig;
  rules?: WafRule[];
};

export type WafThrottleConfig =
  | number
  | {
      name?: string;
      action?: WafAction;
      aggregateKeyType?: 'IP' | 'FORWARDED_IP';
      limit?: number;
      priority?: number;
      forwardedIPConfig?: {
        headerName: string;
        fallbackBehavior: 'MATCH' | 'NO_MATCH';
      };
      scopeDownStatement?: CfnWafRuleStatement;
      visibilityConfig?: VisibilityConfig;
    };

export type WafDisableIntrospectionConfig = {
  name?: string;
  priority?: number;
  visibilityConfig?: VisibilityConfig;
};

export type WafAction = 'Allow' | 'Block';
export type WafRuleAction = 'Allow' | 'Block' | 'Count' | 'Captcha';

export type WafRuleThrottle = {
  throttle: WafThrottleConfig;
};

export type WafRuleCustom = {
  name: string;
  priority?: number;
  action?: WafRuleAction;
  statement: CfnWafRuleStatement;
  visibilityConfig?: VisibilityConfig;
  overrideAction?: Record<string, unknown>;
};

export type WafRuleDisableIntrospection = {
  disableIntrospection: WafDisableIntrospectionConfig;
};

export type WafRule =
  | WafRuleThrottle
  | WafRuleDisableIntrospection
  | WafRuleCustom
  | 'disableIntrospection'
  | 'throttle';

export type ApiKeyConfig = {
  apiKeyId?: string;
  name: string;
  description?: string;
  expiresAfter?: string | number;
  expiresAt?: string;
  wafRules?: WafRule[];
};

export type CognitoAuth = {
  type: 'AMAZON_COGNITO_USER_POOLS';
  config: {
    userPoolId: string | IntrinsicFunction;
    awsRegion?: string | IntrinsicFunction;
    defaultAction?: 'ALLOW' | 'DENY';
    appIdClientRegex?: string | IntrinsicFunction;
  };
};

export type IamAuth = {
  type: 'AWS_IAM';
};

export type LambdaAuth = {
  type: 'AWS_LAMBDA';
  config: LambdaConfig & {
    identityValidationExpression?: string;
    authorizerResultTtlInSeconds?: number;
  };
};

export type OidcAuth = {
  type: 'OPENID_CONNECT';
  config: {
    issuer: string;
    clientId: string;
    iatTTL?: number;
    authTTL?: number;
  };
};

export type ApiKeyAuth = {
  type: 'API_KEY';
};

export type Auth = CognitoAuth | LambdaAuth | OidcAuth | ApiKeyAuth | IamAuth;

export type DomainConfig = {
  enabled?: boolean;
  useCloudFormation?: boolean;
  retain?: boolean;
  name: string;
  certificateArn?: string;
  hostedZoneId?: string;
  hostedZoneName?: string;
  route53?: boolean;
};

export type SyncConfig = {
  conflictDetection: 'VERSION' | 'NONE';
  conflictHandler: 'OPTIMISTIC_CONCURRENCY' | 'AUTOMERGE' | 'LAMBDA';
} & LambdaConfig;

export type Substitutions = Record<string, string | IntrinsicFunction>;
export type EnvironmentVariables = Record<string, string | IntrinsicFunction>;
export type EnhancedMetricsConfig = {
  DataSourceLevelMetricsBehavior:
    | 'FULL_REQUEST_DATA_SOURCE_METRICS'
    | 'PER_DATA_SOURCE_METRICS';
  OperationLevelMetricsConfig: 'ENABLED' | ' DISABLED';
  ResolverLevelMetricsBehavior:
    | 'FULL_REQUEST_RESOLVER_METRICS'
    | 'PER_RESOLVER_METRICS';
};

export type DsDynamoDBConfig = {
  type: 'AMAZON_DYNAMODB';
  config: {
    tableName: string | IntrinsicFunction;
    useCallerCredentials?: boolean;
    serviceRoleArn?: string | IntrinsicFunction;
    region?: string | IntrinsicFunction;
    iamRoleStatements?: IamStatement[];
    versioned?: boolean;
    deltaSyncConfig?: {
      deltaSyncTableName: string;
      baseTableTTL?: number;
      deltaSyncTableTTL?: number;
    };
  };
};

export type DsEventBridgeConfig = {
  type: 'AMAZON_EVENTBRIDGE';
  config: {
    serviceRoleArn?: string | IntrinsicFunction;
    iamRoleStatements?: IamStatement[];
    eventBusArn: string | IntrinsicFunction;
  };
};

export type DsRelationalDbConfig = {
  type: 'RELATIONAL_DATABASE';
  config: {
    region?: string;
    relationalDatabaseSourceType?: 'RDS_HTTP_ENDPOINT';
    serviceRoleArn?: string | IntrinsicFunction;
    dbClusterIdentifier: string | IntrinsicFunction;
    databaseName?: string | IntrinsicFunction;
    schema?: string;
    awsSecretStoreArn: string | IntrinsicFunction;
    iamRoleStatements?: IamStatement[];
  };
};

export type DsOpenSearchConfig = {
  type: 'AMAZON_OPENSEARCH_SERVICE';
  config: {
    domain?: string;
    endpoint?: string | IntrinsicFunction;
    region?: string | IntrinsicFunction;
    serviceRoleArn?: string | IntrinsicFunction;
    iamRoleStatements?: IamStatement[];
  };
};

export type LambdaConfig =
  | {
      functionName: string;
      functionAlias?: string;
    }
  | {
      functionArn: string | IntrinsicFunction;
    }
  | {
      function: Record<string, unknown>;
    };

export type DsLambdaConfig = {
  type: 'AWS_LAMBDA';
  config: {
    serviceRoleArn?: string | IntrinsicFunction;
    iamRoleStatements?: IamStatement[];
  } & LambdaConfig;
};

export type DsHttpConfig = {
  type: 'HTTP';
  config: {
    endpoint: string | IntrinsicFunction;
    serviceRoleArn?: string | IntrinsicFunction;
    iamRoleStatements?: IamStatement[];
    authorizationConfig?: {
      authorizationType: 'AWS_IAM';
      awsIamConfig: {
        signingRegion: string | IntrinsicFunction;
        signingServiceName?: string | IntrinsicFunction;
      };
    };
  };
};

export type DsNone = {
  type: 'NONE';
};

export type VisibilityConfig = {
  name?: string;
  cloudWatchMetricsEnabled?: boolean;
  sampledRequestsEnabled?: boolean;
};

export type LoggingConfig = {
  level: 'ERROR' | 'NONE' | 'ALL' | 'DEBUG' | 'INFO';
  enabled?: boolean;
  excludeVerboseContent?: boolean;
  retentionInDays?: number;
  roleArn?: string | IntrinsicFunction;
};

export type CachingConfig = {
  enabled?: boolean;
  behavior: 'FULL_REQUEST_CACHING' | 'PER_RESOLVER_CACHING';
  type?: string;
  ttl?: number;
  atRestEncryption?: boolean;
  transitEncryption?: boolean;
};
