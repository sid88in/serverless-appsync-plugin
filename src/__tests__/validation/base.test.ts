import { AppSyncConfig } from '../../types';
import { validateConfig } from '../../validation';
import { basicConfig } from '../basicConfig';

describe('Valdiation', () => {
  it('should validate ', () => {
    expect(
      validateConfig({
        ...basicConfig,
        visibility: 'GLOBAL',
        introspection: true,
        queryDepthLimit: 10,
        resolverCountLimit: 10,
        xrayEnabled: true,
        environment: {
          MY_TABLE: 'my-table',
          MY_OTHER_TABLE: { Ref: 'MyOtherTable' },
        },
        tags: {
          foo: 'bar',
        },
        esbuild: {
          target: 'es2020',
          sourcemap: false,
          treeShaking: false,
        },
      }),
    ).toBe(true);

    expect(function () {
      validateConfig({
        visibility: 'FOO',
        introspection: 10,
        queryDepthLimit: 'foo',
        resolverCountLimit: 'bar',
        xrayEnabled: 'BAR',
        unknownPorp: 'foo',
        esbuild: 'bad',
        environment: 'Bad',
      });
    }).toThrowErrorMatchingSnapshot();

    expect(function () {
      validateConfig({
        ...basicConfig,
        queryDepthLimit: 76,
        resolverCountLimit: 1001,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  describe('Log', () => {
    describe('Valid', () => {
      const assertions = [
        {
          name: 'Minimum',
          config: {
            ...basicConfig,
            logging: {
              level: 'ALL',
            },
          } as AppSyncConfig,
        },
        {
          name: 'Full',
          config: {
            ...basicConfig,
            logging: {
              level: 'ALL',
              retentionInDays: 14,
              excludeVerboseContent: true,
              loggingRoleArn: { Ref: 'MyLogGorupArn' },
            },
          } as AppSyncConfig,
        },
      ];

      assertions.forEach((config) => {
        it(`should validate a ${config.name}`, () => {
          expect(validateConfig(config.config)).toBe(true);
        });
      });
    });

    describe('Invalid', () => {
      const assertions = [
        {
          name: 'Invalid',
          config: {
            ...basicConfig,
            logging: {
              level: 'FOO',
              retentionInDays: 'bar',
              excludeVerboseContent: 'buzz',
              loggingRoleArn: 123,
              visibility: 'FOO',
            },
          },
        },
      ];

      assertions.forEach((config) => {
        it(`should validate a ${config.name}`, () => {
          expect(function () {
            validateConfig(config.config);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });
  });

  describe('Waf', () => {
    describe('Valid', () => {
      const assertions = [
        {
          name: 'Minimum',
          config: {
            ...basicConfig,
            waf: {
              rules: [],
            },
          } as AppSyncConfig,
        },
        {
          name: 'Full',
          config: {
            ...basicConfig,
            waf: {
              enabled: true,
              name: 'MyWaf',
              defaultAction: 'Allow',
              description: 'My Waf rules',
              visibilityConfig: {
                name: 'myRule',
                cloudWatchMetricsEnabled: true,
                sampledRequestsEnabled: true,
              },
              rules: [
                'throttle',
                { throttle: 100 },
                {
                  throttle: {
                    name: 'Throttle',
                    action: 'Block',
                    limit: 200,
                    priority: 200,
                    aggregateKeyType: 'IP',
                    forwardedIPConfig: {
                      headerName: 'X-Forwarded-For',
                      fallbackBehavior: 'MATCH',
                    },
                    visibilityConfig: {
                      name: 'throttle200',
                      cloudWatchMetricsEnabled: true,
                      sampledRequestsEnabled: true,
                    },
                  },
                },
                'disableIntrospection',
                {
                  disableIntrospection: {
                    name: 'Disable Intorspection',
                    priority: 100,
                    visibilityConfig: {
                      name: 'DisableIntrospection',
                      cloudWatchMetricsEnabled: true,
                      sampledRequestsEnabled: true,
                    },
                  },
                },
                {
                  name: 'Custom Rule',
                  action: 'Count',
                  priority: 500,
                  statement: {
                    NotStatement: {
                      Statement: {
                        GeoMatchStatement: {
                          CountryCodes: ['US'],
                        },
                      },
                    },
                  },
                  visibilityConfig: {
                    name: 'myRule',
                    cloudWatchMetricsEnabled: true,
                    sampledRequestsEnabled: true,
                  },
                },
              ],
            },
          } as AppSyncConfig,
        },
        {
          name: 'Using arn',
          config: {
            ...basicConfig,
            waf: {
              enabled: true,
              arn: 'arn:aws:',
            },
          },
        },
      ];

      assertions.forEach((config) => {
        it(`should validate a ${config.name}`, () => {
          expect(validateConfig(config.config)).toBe(true);
        });
      });
    });

    describe('Invalid', () => {
      const assertions = [
        {
          name: 'Invalid',
          config: {
            ...basicConfig,
            waf: {
              enabled: 'foo',
              name: 123,
              defaultAction: 'Buzz',
              visibilityConfig: {
                name: 123,
                cloudWatchMetricsEnabled: 456,
                sampledRequestsEnabled: 789,
              },
              rules: [
                'fake',
                { invalid: 100 },
                {
                  name: 123,
                  statement: 456,
                },
              ],
            },
          },
        },
        {
          name: 'Invalid arn',
          config: {
            ...basicConfig,
            waf: {
              arn: 123,
            },
          },
        },
        {
          name: 'Throttle limit',
          config: {
            ...basicConfig,
            waf: {
              rules: [
                { throttle: 99 },
                {
                  throttle: {
                    name: 'Throttle',
                    limit: 99,
                  },
                },
              ],
            },
          },
        },
      ];

      assertions.forEach((config) => {
        it(`should validate a ${config.name}`, () => {
          expect(function () {
            validateConfig(config.config);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });
  });

  describe('Domain', () => {
    describe('Valid', () => {
      const assertions = [
        {
          name: 'Minimum',
          config: {
            ...basicConfig,
            domain: {
              name: 'api.example.com',
              certificateArn: 'arn:aws:',
            },
          } as AppSyncConfig,
        },
        {
          name: 'Full',
          config: {
            ...basicConfig,
            domain: {
              enabled: true,
              certificateArn: 'arn:aws:',
              name: 'api.example.com',
              hostedZoneId: 'Z111111QQQQQQQ',
              hostedZoneName: 'example.com.',
              route53: true,
            },
          } as AppSyncConfig,
        },
        {
          name: 'useCloudFormation: false, missing certificateArn',
          config: {
            ...basicConfig,
            domain: {
              name: 'api.example.com',
              useCloudFormation: false,
            },
          } as AppSyncConfig,
        },
      ];

      assertions.forEach((config) => {
        it(`should validate a ${config.name}`, () => {
          expect(validateConfig(config.config)).toBe(true);
        });
      });
    });

    describe('Invalid', () => {
      const assertions = [
        {
          name: 'Invalid',
          config: {
            ...basicConfig,
            domain: {
              enabled: 'foo',
              name: 'bar',
              certificateArn: 123,
              route53: 123,
            },
          },
        },
        {
          name: 'useCloudFormation: true, certificateArn or hostedZoneId is required',
          config: {
            ...basicConfig,
            domain: {
              name: 'api.example.com',
              useCloudFormation: true,
            },
          },
        },
        {
          name: 'useCloudFormation: not present, certificateArn or hostedZoneId is required',
          config: {
            ...basicConfig,
            domain: {
              name: 'api.example.com',
            },
          },
        },
        {
          name: 'Invalid Route 53',
          config: {
            ...basicConfig,
            domain: {
              name: 'bar',
              certificateArn: 'arn:aws:',
              route53: {
                hostedZoneId: 456,
                hostedZoneName: 789,
              },
            },
          },
        },
      ];

      assertions.forEach((config) => {
        it(`should validate a ${config.name}`, () => {
          expect(function () {
            validateConfig(config.config);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });
  });

  describe('Caching', () => {
    describe('Valid', () => {
      const assertions = [
        {
          name: 'Minimum',
          config: {
            ...basicConfig,
            caching: {
              behavior: 'PER_RESOLVER_CACHING',
            },
          } as AppSyncConfig,
        },
        {
          name: 'Full',
          config: {
            ...basicConfig,
            caching: {
              enabled: true,
              behavior: 'PER_RESOLVER_CACHING',
              type: 'SMALL',
              ttl: 3600,
              atRestEncryption: true,
              transitEncryption: true,
            },
          } as AppSyncConfig,
        },
      ];

      assertions.forEach((config) => {
        it(`should validate a ${config.name}`, () => {
          expect(validateConfig(config.config)).toBe(true);
        });
      });
    });

    describe('Invalid', () => {
      const assertions = [
        {
          name: 'Invalid',
          config: {
            ...basicConfig,
            caching: {
              enabled: 'foo',
              behavior: 'bar',
              type: 'INVALID',
              ttl: 'bizz',
              atRestEncryption: 'bizz',
              transitEncryption: 'bazz',
            },
          },
        },
        {
          name: 'Ttl min value',
          config: {
            ...basicConfig,
            caching: {
              behavior: 'PER_RESOLVER_CACHING',
              ttl: 0,
            },
          },
        },
        {
          name: 'Ttl max value',
          config: {
            ...basicConfig,
            caching: {
              behavior: 'PER_RESOLVER_CACHING',
              ttl: 3601,
            },
          },
        },
      ];

      assertions.forEach((config) => {
        it(`should validate a ${config.name}`, () => {
          expect(function () {
            validateConfig(config.config);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });
  });
});
