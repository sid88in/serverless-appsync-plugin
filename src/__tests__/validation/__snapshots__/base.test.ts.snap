// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Valdiation Caching Invalid should validate a Invalid 1`] = `
"/caching/enabled: must be boolean
/caching/behavior: must be one of 'FULL_REQUEST_CACHING', 'PER_RESOLVER_CACHING'
/caching/type: must be one of 'SMALL', 'MEDIUM', 'LARGE', 'XLARGE', 'LARGE_2X', 'LARGE_4X', 'LARGE_8X', 'LARGE_12X'
/caching/ttl: must be integer
/caching/atRestEncryption: must be boolean
/caching/transitEncryption: must be boolean"
`;

exports[`Valdiation Caching Invalid should validate a Ttl max value 1`] = `"/caching/ttl: must be <= 3600"`;

exports[`Valdiation Caching Invalid should validate a Ttl min value 1`] = `"/caching/ttl: must be >= 1"`;

exports[`Valdiation Domain Invalid should validate a Invalid 1`] = `
"/domain/enabled: must be boolean
/domain/name: must be a valid domain name
/domain/certificateArn: must be a string or a CloudFormation intrinsic function
/domain/route53: must be boolean"
`;

exports[`Valdiation Domain Invalid should validate a Invalid Route 53 1`] = `
"/domain/name: must be a valid domain name
/domain/route53: must be boolean"
`;

exports[`Valdiation Domain Invalid should validate a useCloudFormation: not present, certificateArn or hostedZoneId is required 1`] = `"/domain: when using CloudFormation, you must provide either certificateArn or hostedZoneId."`;

exports[`Valdiation Domain Invalid should validate a useCloudFormation: true, certificateArn or hostedZoneId is required 1`] = `"/domain: when using CloudFormation, you must provide either certificateArn or hostedZoneId."`;

exports[`Valdiation Log Invalid should validate a Invalid 1`] = `
"/logging/level: must be one of 'ALL', 'INFO', 'DEBUG', 'ERROR' or 'NONE'
/logging/retentionInDays: must be integer
/logging/excludeVerboseContent: must be boolean"
`;

exports[`Valdiation Waf Invalid should validate a Invalid 1`] = `
"/waf/name: must be string
/waf/defaultAction: must be 'Allow' or 'Block'
/waf/rules/0: must be a valid WAF rule
/waf/rules/1: must be a valid WAF rule
/waf/rules/2: must be a valid WAF rule
/waf/enabled: must be boolean"
`;

exports[`Valdiation Waf Invalid should validate a Invalid arn 1`] = `"/waf/arn: must be a string or a CloudFormation intrinsic function"`;

exports[`Valdiation Waf Invalid should validate a Throttle limit 1`] = `
"/waf/rules/0: must be a valid WAF rule
/waf/rules/1: must be a valid WAF rule"
`;

exports[`Valdiation should validate  1`] = `
": must have required property 'name'
: must have required property 'authentication'
/unknownPorp: invalid (unknown) property
/xrayEnabled: must be boolean
/visibility: must be \\"GLOBAL\\" or \\"PRIVATE\\"
/introspection: must be boolean
/queryDepthLimit: must be integer
/resolverCountLimit: must be integer
/environment: must be a valid environment definition
/esbuild: must be an esbuild config object or false"
`;

exports[`Valdiation should validate  2`] = `
"/queryDepthLimit: must be <= 75
/resolverCountLimit: must be <= 1000"
`;
